python.d/boinc.chart.py

```diff
@@ -0,0 +1,178 @@
+    'comperror': 0,
+    'uploading': 0,
+    'uploaded': 0,
+    'aborted': 0,
+    'upload_failed': 0,
+    'uninit_sched': 0,
+    'preempted': 0,
+    'scheduled': 0,
+    'uninit_proc': 0,
+    'executing': 0,
+    'suspended': 0,
+    'aborting': 0,
+    'quit': 0,
+    'copy_pending': 0
+}
+
+# Map task states to dimensions
+_TASK_MAP = {
+    boinc_client.ResultState.NEW: 'new',
+    boinc_client.ResultState.FILES_DOWNLOADING: 'downloading',
+    boinc_client.ResultState.FILES_DOWNLOADED: 'downloaded',
+    boinc_client.ResultState.COMPUTE_ERROR: 'comperror',
+    boinc_client.ResultState.FILES_UPLOADING: 'uploading',
+    boinc_client.ResultState.FILES_UPLOADED: 'uploaded',
+    boinc_client.ResultState.ABORTED: 'aborted',
+    boinc_client.ResultState.UPLOAD_FAILED: 'upload_failed'
+}
+
+# Map scheduler states to dimensions
+_SCHED_MAP = {
+    boinc_client.CpuSched.UNINITIALIZED: 'uninit_sched',
+    boinc_client.CpuSched.PREEMPTED: 'preempted',
+    boinc_client.CpuSched.SCHEDULED: 'scheduled',
+}
+
+# Maps process states to dimensions
+_PROC_MAP = {
+    boinc_client.Process.UNINITIALIZED: 'uninit_proc',
+    boinc_client.Process.EXECUTING: 'executing',
+    boinc_client.Process.SUSPENDED: 'suspended',
+    boinc_client.Process.ABORT_PENDING: 'aborted',
+    boinc_client.Process.QUIT_PENDING: 'quit',
+    boinc_client.Process.COPY_PENDING: 'copy_pending'
+}
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+        self.host = self.configuration.get('host', 'localhost')
+        self.port = self.configuration.get('port', 0)
+        self.password = self.configuration.get('password', '')
+        self.client = boinc_client.BoincClient(host=self.host, port=self.port, passwd=self.password)
+        self.alive = False
+
+    def check(self):
+        if platform.system() != 'Linux':
+            self.error('Only supported on Linux.')
+            return False
+        self.connect()
+        if self.client.connected and self.client.authorized:
+            self.alive = True
+        else:
+            self.alive = False
+        return self.alive
+
+    def connect(self):
+        self.client.connect()
+
+    def reconnect(self):
+        try:
+            self.client.disconnect()
+        except socket.error:
+            pass
+        self.client.connect()
+        if self.client.connected and self.client.authorized:
+            self.alive = True
+        else:
+            self.alive = False
+        return self.alive
+
+    def is_alive(self):
+        if (not self.alive) or \
+           self.client.rpc.sock.getsockopt(socket.IPPROTO_TCP, socket.TCP_INFO, 0) != 1:
+            return self.reconnect()
+        return True
+
+    def _get_data(self):
+        if not self.is_alive():
+            return None
+        data = deepcopy(_DATA_TEMPLATE)
```

Comment:
`_DATA_TEMPLATE` is one level `dict` so 
```python
dict(_DATA_TEMPLATE)
#  or _DATA_TEMPLATE.copy() 
```
is enough

Reply:
Good point.  That first one should be significantly more efficient too.